src/vs/workbench/contrib/positronNotebook/browser/PositronNotebookInstance.ts

```diff
@@ -975,6 +992,144 @@ export class PositronNotebookInstance extends Disposable implements IPositronNot
+	copyCells(cells?: IPositronNotebookCell[]): void {
+		const cellsToCopy = cells || this.getSelectedCells();
+
+		if (cellsToCopy.length === 0) {
+			return;
+		}
+
+		// Store internally for full-fidelity paste
+		this._clipboardCells = cellsToCopy.map(cell => cellToCellDto2(cell));
+		this._isClipboardCut = false;
+
+		// Also write to system clipboard as text
+		const clipboardText = serializeCellsToClipboard(cellsToCopy);
+		this._clipboardService.writeText(clipboardText);
+
+		// Log for debugging
+		this._logService.debug(`Copied ${cellsToCopy.length} cells to clipboard`);
+	}
+
+	cutCells(cells?: IPositronNotebookCell[]): void {
+		const cellsToCut = cells || this.getSelectedCells();
+
+		if (cellsToCut.length === 0) {
+			return;
+		}
+
+		// Copy cells first
+		this.copyCells(cellsToCut);
+		this._isClipboardCut = true;
+
+		// Delete the cells (this handles selection and focus automatically)
+		this.deleteCells(cellsToCut);
+	}
+
+	pasteCells(index?: number): void {
+		if (!this.canPaste()) {
+			return;
+		}
+
+		this._assertTextModel();
+
+		const pasteIndex = index ?? this.getInsertionIndex();
+		const cellCount = this._clipboardCells.length;
+
+		// Use textModel.applyEdits to properly create and register cells
+		const synchronous = true;
+		const pushUndoStop = true;
+		const endSelections: ISelectionState = {
+			kind: SelectionStateType.Index,
+			focus: { start: pasteIndex, end: pasteIndex + cellCount },
+			selections: [{ start: pasteIndex, end: pasteIndex + cellCount }]
+		};
+		const focusAfterInsertion = {
+			start: pasteIndex,
+			end: pasteIndex + cellCount
+		};
+
+		this.textModel.applyEdits([
+			{
+				editType: CellEditType.Replace,
+				index: pasteIndex,
+				count: 0,
+				cells: this._clipboardCells
+			}
+		],
+			synchronous,
+			{
+				kind: SelectionStateType.Index,
+				focus: focusAfterInsertion,
+				selections: [focusAfterInsertion]
+			},
+			() => endSelections, undefined, pushUndoStop && !this.isReadOnly
+		);
+
+		// If this was a cut operation, clear the clipboard
+		if (this._isClipboardCut) {
+			this._clipboardCells = [];
+			this._isClipboardCut = false;
+		}
+
+		this._onDidChangeContent.fire();
+	}
+
+	pasteCellsAbove(): void {
+		const selection = this.getSelectedCells();
+		if (selection.length > 0) {
+			const firstSelectedIndex = this.cells.get().indexOf(selection[0]);
+			this.pasteCells(firstSelectedIndex);
+		} else {
+			this.pasteCells(0);
+		}
+	}
+
+	canPaste(): boolean {
+		return this._clipboardCells.length > 0;
+	}
+
+	// Helper method to get selected cells
+	private getSelectedCells(): IPositronNotebookCell[] {
+		const state = this.selectionStateMachine.state.get();
+
+		switch (state.type) {
+			case SelectionState.SingleSelection:
+			case SelectionState.MultiSelection:
+				return state.selected;
+			case SelectionState.EditingSelection:
+				return [state.selectedCell];
+			case SelectionState.NoSelection:
+			default:
+				return [];
+		}
```

Comment:
I've seen something like this in a few places. Perhaps this could live in a `get selectedCells` helper on the `selectionStateMachine.state` object?

Reply:
That does make a lot more sense!